_posts/2023-01-06-Algorithms-Graphs-BFS-DFS.md

```diff
@@ -0,0 +1,344 @@
+
+Having this knowledge let's implement DFS and BFS in Swift. Additionally let's take a look at some additional topics which are very interesting as well:
+
+- Finding the shortest path with unweighted graph
+- Finding a cycle in an undirected graph
+- Finding a cycle in a directed graph
+
+#### DFS
+
+```swift
+var visited: Set<Int> = []
+
+func dfs(_ root: TreeNode?)  {
+
+	guard let root = root else {
+		return
+	}
+
+	visited.insert(root.val)
+
+	for child in root.children {
+		if !visited.contains(child.val) {
+			dfs(child)
+		}
+	}
+}
+```
+
+#### BFS
+
+```swift
+
+class TreeNode {
+	let id: Int
+	let children: [TreNode]
+	
+	init(_ id: Int, _ children: [TreNode] = []) {
+		self.id = id
+		self.children = children
+	}
+}
+
+func bfs(_ start: Node) {
+
+	var queue: [Node] = []
+	var visited: Set<Node> = []
+
+	queue.append(start)
+
+	while !queue.isEmpty {
+  		for item in queue {
+  		let first = queue.removeFirst()
+   			for child in first.children {
+				if !visited.contains(child) {
+					visited.insert(child)
+					queue.append(child)
+				}
+			}
+		}
+	}
+}	
+```
+
+#### BFS + shortest path
+
+Implementation with additional finding distance from the start node and printing a path (an important note here is that here we are discussing unweighted graphs. If we want to find the shortest distance for a weighted graph, we need to consider another algorithm like [Dijktstra's algorithm](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm)).
+
+
+```swift
+func bfs(_ start: Node) {
+	var queue: [Node] = []
+	var visited: Set<Node> = []
+	var distance: [Node: Int] = [:]
+	var parent: [Node: Node] = [:]
+
+	queue.append(start)
+
+	while !queue.isEmpty {
+		for item in queue {
+			let first = queue.removeFirst()
+			for child in first.children {
+				if !visited.contains(child) {
+					dist[child] = dist[item] + 1
+					parent[child] = item 
+  					visited.insert(child)
+					queue.append(child)
+				}
+			}
+		}
+	}
+}
+
+func getPath(_ target: Node, _ start: Node) {
+  var x = target
+  var path: [Int] = []
+  
+  path.append(target)
+  
+  while x != start {
+    x = parent[x]
+    path.append(x)
+  }
+  
+  path.reverse()
+}
+```
+
+
+#### Detecting cycle in undirected graph
+
+```swift
+
+var visited: Set<Node> = []
+
+func dfs(_ root: TreeNode?, stack: [Int]) {	
+
+	guard let root = root else {
+		return
+  	}
+
+	visited.insert(root)
+	let children = root.children
+	for child in children {
+		if !visited.contains(child) {
+			dfs(child, stack + child)
+		} else if child != previous {
+    		// here we can be sure that we don't have a cycle consists of only two elements
+    		// and then we can print stack starting from child
+		}
+	}
+}
+```
+
+#### Detecting cycle in a directed graph
+
+Undirected graphs are much more easier to handle, because we don't care about direction of each edge. In this section I would like to show you how to detect the cycle in a directed graph and as well how to find a path for detected cycle. 
```

Comment:
```suggestion
Undirected graphs are much easier to handle, because we don't care about direction of each edge. In this section I would like to show you how to detect the cycle in a directed graph and as well how to find a path for detected cycle. 
```